src/Microsoft.Identity.Web/TokenCacheProviders/Session/SessionTokenCacheProviderExtension.cs

```diff
@@ -172,11 +194,20 @@ public static IServiceCollection AddSessionPerUserTokenCache(this IServiceCollec
 
             services.AddHttpContextAccessor();
             services.AddSession(option =>
-                {
-                    option.Cookie.IsEssential = true;
-                });
+            {
+                option.Cookie.IsEssential = true;
+            });
             services.AddScoped<IMsalTokenCacheProvider, MsalSessionTokenCacheProvider>();
-            services.TryAddScoped(provider => provider.GetService<IHttpContextAccessor>().HttpContext.Session);
+            services.TryAddScoped(provider =>
```

Comment:
This won't work with Graph which is a singleton. Is there a possible mitigation?

Reply:
> Because session token caches are added with scoped lifetime, they should not be used when `TokenAcquisition` is also used as a singleton (for example, when using Microsoft Graph SDK).

I added the above note to each method and to the [Token cache serialization](https://github.com/AzureAD/microsoft-identity-web/wiki/Token-cache-serialization) wiki.

In this method, I don't think we can reliably predict which services are added, since they can be modified in any order.

Maybe if `ServiceProvider` has some sort of validator or event that we can plug into, we could validate at that time.